ktoml-core/src/commonMain/kotlin/com/akuleshov7/ktoml/decoders/TomlAbstractDecoder.kt

```diff
@@ -49,25 +51,62 @@ public abstract class TomlAbstractDecoder : AbstractDecoder() {
 
     internal abstract fun decodeKeyValue(): TomlKeyValue
 
-    private fun invalidType(typeName: String, requiredType: String): Nothing {
-        val keyValue = decodeKeyValue()
-        throw IllegalTypeException(
-            "<$typeName> type is not allowed by toml specification," +
-                    " use <$requiredType> instead" +
-                    " (key = ${keyValue.key.content}; value = ${keyValue.value.content})", keyValue.lineNo
-        )
-    }
-
+    /**
+     * This is just an adapter from `kotlinx.serialization` to match the content with a type from a Toml Tree,
+     * that we have parsed to a type that is described in user's code. For example:
+     * >>> input: a = "5"
+     * >>> stored in Toml Tree: TomlString("5")
+     * >>> expected by user: data class A(val a: Int)
+     * >>> TomlString cannot be cast to Int, user made a mistake -> IllegalTypeException
+     */
     private inline fun <reified T> decodePrimitiveType(): T {
         val keyValue = decodeKeyValue()
         try {
-            return keyValue.value.content as T
+            return when (val value = keyValue.value) {
+                is TomlLong -> decodeInteger(value.content as Long, keyValue.lineNo)
+                else -> keyValue.value.content as T
+            }
         } catch (e: ClassCastException) {
-            throw CastException(
+            throw IllegalTypeException(
                 "Cannot decode the key [${keyValue.key.content}] with the value [${keyValue.value.content}]" +
                         " with the provided type [${T::class}]. Please check the type in your Serializable class or it's nullability",
                 keyValue.lineNo
             )
         }
     }
+
+    /**
+     * After a lot of discussions, we have finally decided to allow to use Integer types and not only Long.
```

Comment:
> After a lot of discussions

I think it would be valuable to add a link to the discussions, so there's some context